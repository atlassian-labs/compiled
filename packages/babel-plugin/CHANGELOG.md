# @compiled/babel-plugin

## 0.12.0

### Minor Changes

- f139218: Handle destructuring `property: value` pairs returned from arrow functions, and add support for nested and alias destructuring.
- 858146c: Add babel plugins support
- b0adb8a: Added support for conditional expressions when passing an array to the `css` prop of an element

### Patch Changes

- b0adb8a: Fix support for CSS helper call expressions when used in conditional expressions (i.e. the `css(...)` function provided by compiled)

## 0.11.4

### Patch Changes

- 254a6f6: Added ESLint rule to prevent use of extraneous packages, and added these usages of these packages as dependencies. Added new namespace `@compiled-private` to prevent name clashes with existing npm packages.
- c757259: Update type definition dependencies
- 63148ec: Support file importing in babel plugin and add configuration in loaders
- Updated dependencies [c757259]
- Updated dependencies [63148ec]
  - @compiled/css@0.8.1

## 0.11.3

### Patch Changes

- 3b7c188: Refactor the way member expressions are statically evaluated and handle more combinations of expressions within a member
- c2ae4eb: Resolve css-what and nth-check to new versions in @compiled/css
- Updated dependencies [c2ae4eb]
- Updated dependencies [b345bf4]
  - @compiled/css@0.8.0
  - @compiled/utils@0.6.14

## 0.11.2

### Patch Changes

- 8c9ab8c: Update `homepage` and other `package.json` properties
- Updated dependencies [8c9ab8c]
- Updated dependencies [1ca19be]
  - @compiled/css@0.7.2
  - @compiled/utils@0.6.13

## 0.11.1

### Patch Changes

- 79cfb08: Compiled will no longer try to traverse modules boundaries of its own.
- 14368bb: Fix issue where a styled value function using both object destructuring and a template literal in at least one branch resulted in a CSS error
- 68ebac3: Add support for namespace imports and export specifiers
- 427cead: Compiled now supports turning on the `css` prop using jsx pragmas (both with `@jsx` and `@jsxImportSource`).
- 79cfb08: Internal refactor changing how the TypeScript compiler picks up source files.
- Updated dependencies [79cfb08]
  - @compiled/css@0.7.1
  - @compiled/utils@0.6.12

## 0.11.0

### Minor Changes

- fa6af90: Add support for nested ternary operators. Additionally, Compiled will no longer transform ternaries into logical statements unless one side is undefined.

### Patch Changes

- e015a3a: Add comment directive `// @compiled-disable(-next)-line) transform-css-prop` to disable Compiled processing on CSS prop

## 0.10.0

### Minor Changes

- 53a3d71: **Breaking change:** Ternary conditionals will no longer add falsy path styles when expression evaluates truthy

### Patch Changes

- b68411c: Fix styled path check

## 0.9.0

### Minor Changes

- 0b60ae1: Support custom `resolver`
- 2092839: Allow inline strings and inline css mixins in conditional expressions. Fix ordering of styles in template literals.

## 0.8.0

### Minor Changes

- 53935b3: Add `ObjectExpression` support to `css`

## 0.7.0

### Minor Changes

- bcb2a68: Add support for `keyframes`
- a7ab8e1: Add support for conditional rules for `Styled`

### Patch Changes

- e1dc346: Fix missing key prop on generated React elements
- 48805ec: Use the correct expression in the style prop, when an identifier is shadowed by a function argument
- 587e729: Generate CSS for rules defined before a mixin and ensure that mixins can be overriden
- Updated dependencies [bcb2a68]
  - @compiled/css@0.7.0

## 0.6.14

### Patch Changes

- 30ddaf4: Adds a support of computed properties static evaluation in object styles.

## 0.6.13

### Patch Changes

- 40bc0d9: Package descriptions have been updated.
- 1b1c964: The `css` mixin API is now available,
  functioning similarly to the [emotion equivalent](https://emotion.sh/docs/composition).

  ```jsx
  import { css } from '@compiled/react';

  <div
    css={css`
      display: flex;
      font-size: 50px;
      color: blue;
    `}
  >
    blue text
  </div>;
  ```

  For more help, read the docs: https://compiledcssinjs.com/docs/css.

- Updated dependencies [40bc0d9]
- Updated dependencies [1b1c964]
  - @compiled/css@0.6.11
  - @compiled/utils@0.6.11

## 0.6.12

### Patch Changes

- ca573d7: Styled APIs now have display names when running development builds.

## 0.6.10

### Patch Changes

- 37108e4: Compiled dependencies are now using carat range.
- Updated dependencies [992e401]
- Updated dependencies [37108e4]
  - @compiled/utils@0.6.10
  - @compiled/css@0.6.10

## 0.6.9

### Patch Changes

- Updated dependencies [0bb1c11]
- Updated dependencies [0bb1c11]
  - @compiled/css@0.6.9
  - @compiled/utils@0.6.9

## 0.6.8

### Patch Changes

- aea3504: Packages now released with [changesets](https://github.com/atlassian/changesets).
- Updated dependencies [aea3504]
  - @compiled/css@0.6.8
  - @compiled/utils@0.6.8
