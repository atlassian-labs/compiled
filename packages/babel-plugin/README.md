# @compiled/babel-plugin

Babel plugin used to transform your code.

## Installation

```bash
npm i @compiled/babel-plugin
```

## Usage

Detailed docs and example usage can be [found on the documentation website](https://compiledcssinjs.com/docs/pkg-babel-plugin).
