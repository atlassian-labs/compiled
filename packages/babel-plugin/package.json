{
  "name": "@compiled/babel-plugin",
  "version": "0.4.7",
  "description": "The CSS in JS authoring experience you love without the runtime cost",
  "author": "Michael Dougall",
  "license": "Apache-2.0",
  "homepage": "https://compiledcssinjs.com",
  "bugs": "https://github.com/atlassian-labs/compiled-css-in-js/issues/new?assignees=&labels=bug&template=bug_report.md",
  "repository": {
    "type": "git",
    "url": "https://github.com/atlassian-labs/compiled-css-in-js.git",
    "directory": "packages/babel-plugin"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "dependencies": {
    "@compiled/css": "0.4.7",
    "@babel/core": "^7.10.2",
    "@babel/generator": "^7.10.2",
    "@babel/helper-plugin-utils": "^7.10.2",
    "@babel/plugin-syntax-jsx": "^7.10.2",
    "@babel/plugin-transform-flow-strip-types": "^7.10.2",
    "@babel/template": "^7.10.2",
    "@babel/traverse": "^7.10.2",
    "@babel/types": "^7.10.2",
    "@emotion/is-prop-valid": "^0.8.6"
  },
  "devDependencies": {
    "@types/babel__core": "^7.1.7"
  }
}
