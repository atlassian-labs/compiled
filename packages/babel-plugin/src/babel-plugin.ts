import { basename, resolve, join, dirname } from 'path';

import { declare } from '@babel/helper-plugin-utils';
import jsxSyntax from '@babel/plugin-syntax-jsx';
import template from '@babel/template';
import type { NodePath, Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import { unique, preserveLeadingComments, JSX_ANNOTATION_REGEX } from '@compiled/utils';

import { visitClassNamesPath } from './class-names';
import { visitCssMapPath } from './css-map';
import { visitCssPropPath } from './css-prop';
import { visitStyledPath } from './styled';
import type { State } from './types';
import { appendRuntimeImports } from './utils/append-runtime-imports';
import { buildCodeFrameError } from './utils/ast';
import { Cache } from './utils/cache';
import {
  isCompiledCSSCallExpression,
  isCompiledCSSTaggedTemplateExpression,
  isCompiledKeyframesCallExpression,
  isCompiledKeyframesTaggedTemplateExpression,
  isCompiledStyledCallExpression,
  isCompiledStyledTaggedTemplateExpression,
  isCompiledCSSMapCallExpression,
} from './utils/is-compiled';
import { isTransformedJsxFunction } from './utils/is-jsx-function';
import { normalizePropsUsage } from './utils/normalize-props-usage';
import { visitXcssPropPath } from './xcss-prop';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require('../package.json');
const JSX_SOURCE_ANNOTATION_REGEX = /\*?\s*@jsxImportSource\s+([^\s]+)/;
const DEFAULT_IMPORT_SOURCES = ['@compiled/react', '@atlaskit/css'];

let globalCache: Cache | undefined;

const findClassicJsxPragmaImport: Visitor<State> = {
  ImportSpecifier(path, state) {
    const specifier = path.node;

    t.assertImportDeclaration(path.parent);
    // We don't care about other libraries
    if (!this.importSources.includes(path.parent.source.value)) return;

    if (
      (specifier.imported.type === 'StringLiteral' && specifier.imported.value === 'jsx') ||
      (specifier.imported.type === 'Identifier' && specifier.imported.name === 'jsx')
    ) {
      // Hurrah, we know that the jsx function in the JSX pragma refers to the
      // jsx function from Compiled.
      state.pragma.classicJsxPragmaIsCompiled = true;
      state.pragma.classicJsxPragmaLocalName = specifier.local.name;

      // Remove the jsx import; the assumption is that we removed the classic JSX pragma, so
      // Babel shouldn't convert React.createElement to the jsx function anymore.
      path.remove();
      return;
    }
  },
};

export default declare<State>((api) => {
  api.assertVersion(7);

  return {
    name: packageJson.name,
    inherits: jsxSyntax,
    pre(state) {
      const rootPath = state.opts.root ?? this.cwd;

      this.sheets = {};
      this.cssMap = {};
      let cache: Cache;

      if (this.opts.cache === true) {
        globalCache = new Cache();
        cache = globalCache;
      } else {
        cache = new Cache();
      }

      cache.initialize({ ...this.opts, cache: !!this.opts.cache });

      this.cache = cache;
      this.includedFiles = [];
      this.pathsToCleanup = [];
      this.pragma = {};
      this.usesXcss = false;
      this.importSources = [
        ...DEFAULT_IMPORT_SOURCES,
        ...(this.opts.importSources
          ? this.opts.importSources.map((origin) => {
              if (origin[0] === '.') {
                // We've found a relative path, transform it to be fully qualified.
                return join(rootPath, origin);
              }

              return origin;
            })
          : []),
      ];

      if (typeof this.opts.resolver === 'object') {
        this.resolver = this.opts.resolver;
      } else if (typeof this.opts.resolver === 'string') {
        this.resolver = require(require.resolve(this.opts.resolver, {
          paths: [rootPath],
        }));
      }

      this.transformCache = new WeakMap();
    },
    visitor: {
      Program: {
        enter(path, state) {
          const { file } = state;
          let jsxComment: t.Comment | undefined;

          // Handle classic JSX pragma, if it exists
          path.traverse<State>(findClassicJsxPragmaImport, this);

          if (!file.ast.comments) {
            return;
          }

          for (const comment of file.ast.comments) {
            const jsxSourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value);
            const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);

            // jsxPragmas currently only run on the top-level compiled module,
            // hence we don't interrogate this.importSources.
            if (jsxSourceMatches && this.importSources.includes(jsxSourceMatches[1])) {
              // jsxImportSource pragma found - turn on CSS prop!
              state.compiledImports = {};
              state.pragma.jsxImportSource = true;
              jsxComment = comment;
            }

            if (
              jsxMatches &&
              state.pragma.classicJsxPragmaIsCompiled &&
              jsxMatches[1] === state.pragma.classicJsxPragmaLocalName
            ) {
              state.compiledImports = {};
              state.pragma.jsx = true;
              jsxComment = comment;
            }
          }

          if (jsxComment) {
            // Delete the JSX pragma from the file, so that JSX
            // elements don't get converted to jsx functions when using Compiled.
            // This is to avoid having an import from a library that isn't
            // `@compiled/react/runtime` in the final output:
            //
            //     import { jsx } from '@compiled/react'
            //     import { jsx as _jsx } from '@compiled/react/jsx-runtime';
            //     import { jsxs as _jsxs } from '@compiled/react/jsx-runtime';

            // Hide the JSX pragma from the
            // @babel/plugin-transform-react-jsx plugin
            file.ast.comments = file.ast.comments.filter((c: t.Comment) => c !== jsxComment);

            // Remove the JSX pragma comment from
            // the Babel output.
            //
            // Note that Babel provides no way for us to traverse comments >:(
            // So the best we can do is guess that the JSX pragma is probably at the start of the file.
            if (path.node.body[0].leadingComments) {
              path.node.body[0].leadingComments = path.node.body[0].leadingComments.filter(
                (newComment) => newComment !== jsxComment
              );
            }
          }
        },
        exit(path, state) {
          if (!state.compiledImports && !state.usesXcss) {
            return;
          }

          const { pragma } = state;

          // Always import React if the developer is using
          // /** @jsx jsx */, because these will get converted
          // to React.createElement function calls
          const shouldImportReact = state.pragma.jsx || (state.opts.importReact ?? true);

          preserveLeadingComments(path);

          appendRuntimeImports(path, state);

          if (!pragma.jsxImportSource && shouldImportReact && !path.scope.getBinding('React')) {
            // React is missing - add it in at the last moment!
            path.unshiftContainer('body', template.ast(`import * as React from 'react'`));
          }

          if (state.compiledImports?.styled && !path.scope.getBinding('forwardRef')) {
            // forwardRef is missing - add it in at the last moment!
            path.unshiftContainer('body', template.ast(`import { forwardRef } from 'react'`));
          }

          const filename = basename(state.filename ?? '') || 'File';
          const version = process.env.TEST_PKG_VERSION || packageJson.version;

          path.addComment('leading', ` ${filename} generated by ${packageJson.name} v${version} `);

          // Add a line break after the comment
          path.unshiftContainer('body', t.noop());

          // Callback when included files have been added.
          if (this.includedFiles.length && this.opts.onIncludedFiles) {
            this.opts.onIncludedFiles(unique(this.includedFiles));
          }

          // Cleanup paths that have been marked.
          state.pathsToCleanup.forEach((clean) => {
            switch (clean.action) {
              case 'remove': {
                clean.path.remove();
                return;
              }

              case 'replace': {
                clean.path.replaceWith(t.nullLiteral());
                return;
              }

              default:
                return;
            }
          });
        },
      },
      ImportDeclaration(path, state) {
        const userLandModule = path.node.source.value;

        const isCompiledModule = this.importSources.some((compiledModuleOrigin) => {
          if (compiledModuleOrigin === userLandModule) {
            return true;
          }

          if (
            state.filename &&
            userLandModule[0] === '.' &&
            userLandModule.endsWith(basename(compiledModuleOrigin))
          ) {
            // Relative import that might be a match, resolve the relative path and compare.
            const fullpath = resolve(dirname(state.filename), userLandModule);
            return fullpath === compiledModuleOrigin;
          }

          return false;
        });

        if (!isCompiledModule) {
          return;
        }

        // The presence of the module enables CSS prop
        state.compiledImports = state.compiledImports || {};

        // Go through each import and enable each found API
        path.get('specifiers').forEach((specifier) => {
          if (!state.compiledImports || !specifier.isImportSpecifier()) {
            // Bail out early
            return;
          }

          (['styled', 'ClassNames', 'css', 'keyframes', 'cssMap'] as const).forEach((apiName) => {
            if (
              state.compiledImports &&
              t.isIdentifier(specifier.node?.imported) &&
              specifier.node?.imported.name === apiName
            ) {
              // Enable the API with the local name
              const apiArray = state.compiledImports[apiName] || [];
              apiArray.push(specifier.node.local.name);
              state.compiledImports[apiName] = apiArray;
              specifier.remove();
            }
          });
        });

        if (path.node.specifiers.length === 0) {
          path.remove();
        }
      },
      'TaggedTemplateExpression|CallExpression'(
        path: NodePath<t.TaggedTemplateExpression> | NodePath<t.CallExpression>,
        state: State
      ) {
        if (isTransformedJsxFunction(path, state)) {
          throw buildCodeFrameError(
            `Found a \`jsx\` function call in the Babel output where one should not have been generated. Was Compiled not set up correctly?

Reasons this might happen:

[Likely] Importing \`jsx\` from a library other than Compiled CSS-in-JS - please only import from \`@compiled/react\`.

[Less likely] If you are using \`@babel/preset-react\` (or \`@babel/plugin-transform-react-jsx\`) in your Babel configuration, and you are using \`runtime: classic\`, make sure you do not use the \`pragma\` option. Please use the /** @jsx jsx */ syntax instead, or switch to \`runtime: automatic\``,
            // Use parent node to mitigate likelihood of
            // "This is an error on an internal node." warning in the
            // error output
            path.parentPath.node,
            path.parentPath
          );
        }
        if (isCompiledCSSMapCallExpression(path.node, state)) {
          visitCssMapPath(path, { context: 'root', state, parentPath: path });
          return;
        }

        const hasStyles =
          isCompiledCSSTaggedTemplateExpression(path.node, state) ||
          isCompiledStyledTaggedTemplateExpression(path.node, state) ||
          isCompiledCSSCallExpression(path.node, state) ||
          isCompiledStyledCallExpression(path.node, state);

        if (hasStyles) {
          normalizePropsUsage(path);
        }

        const isCompiledUtil =
          isCompiledCSSTaggedTemplateExpression(path.node, state) ||
          isCompiledKeyframesTaggedTemplateExpression(path.node, state) ||
          isCompiledCSSCallExpression(path.node, state) ||
          isCompiledKeyframesCallExpression(path.node, state);

        if (isCompiledUtil) {
          state.pathsToCleanup.push({ path, action: 'replace' });
          return;
        }

        const isCompiledComponent =
          isCompiledStyledTaggedTemplateExpression(path.node, state) ||
          isCompiledStyledCallExpression(path.node, state);

        if (isCompiledComponent) {
          visitStyledPath(path, { context: 'root', state, parentPath: path });
          return;
        }
      },
      JSXElement(path, state) {
        if (!state.compiledImports?.ClassNames) {
          return;
        }

        visitClassNamesPath(path, { context: 'root', state, parentPath: path });
      },
      JSXOpeningElement(path, state) {
        const compiledXCSSProp = state.opts.processXcss ?? true;
        if (compiledXCSSProp) {
          visitXcssPropPath(path, { context: 'root', state, parentPath: path });
        }

        if (state.compiledImports) {
          visitCssPropPath(path, { context: 'root', state, parentPath: path });
        }
      },
    },
  };
});
