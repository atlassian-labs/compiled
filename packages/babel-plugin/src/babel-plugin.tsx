import { basename } from 'path';

import { declare } from '@babel/helper-plugin-utils';
import jsxSyntax from '@babel/plugin-syntax-jsx';
import template from '@babel/template';
import * as t from '@babel/types';
import { unique } from '@compiled/utils';

import { visitClassNamesPath } from './class-names';
import { visitCssPropPath } from './css-prop';
import { visitStyledPath } from './styled';
import type { State } from './types';
import { appendRuntimeImports } from './utils/append-runtime-imports';
import { Cache } from './utils/cache';
import {
  isCompiledCSSCallExpression,
  isCompiledCSSTaggedTemplateExpression,
  isCompiledKeyframesCallExpression,
  isCompiledKeyframesTaggedTemplateExpression,
  isCompiledStyledCallExpression,
  isCompiledStyledTaggedTemplateExpression,
} from './utils/is-compiled';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require('../package.json');
const JSX_SOURCE_ANNOTATION_REGEX = /\*?\s*@jsxImportSource\s+([^\s]+)/;
const JSX_ANNOTATION_REGEX = /\*?\s*@jsx\s+([^\s]+)/;
const COMPILED_MODULE = '@compiled/react';

let globalCache: Cache | undefined;

export default declare<State>((api) => {
  api.assertVersion(7);

  return {
    name: packageJson.name,
    inherits: jsxSyntax,
    pre() {
      this.sheets = {};
      let cache: Cache;

      if (this.opts.cache === true) {
        globalCache = new Cache();
        cache = globalCache;
      } else {
        cache = new Cache();
      }

      cache.initialize({ ...this.opts, cache: !!this.opts.cache });

      this.cache = cache;
      this.includedFiles = [];
      this.pathsToCleanup = [];
      this.pragma = {};
    },
    visitor: {
      Program: {
        enter(_, state) {
          const { file } = state;

          if (file.ast.comments) {
            for (const comment of file.ast.comments) {
              const jsxSourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value);
              const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);

              if (jsxSourceMatches && jsxSourceMatches[1] === COMPILED_MODULE) {
                // jsxImportSource pragma found - turn on CSS prop!
                state.compiledImports = {};
                state.pragma.jsxImportSource = true;
              }

              if (jsxMatches && jsxMatches[1] === 'jsx') {
                state.pragma.jsx = true;
              }
            }
          }
        },
        exit(path, state) {
          if (!state.compiledImports) {
            return;
          }

          const {
            pragma,
            opts: { importReact: shouldImportReact = true },
          } = state;

          appendRuntimeImports(path);

          const hasPragma = pragma.jsxImportSource || pragma.jsx;

          if (!hasPragma && shouldImportReact && !path.scope.getBinding('React')) {
            // React is missing - add it in at the last moment!
            path.unshiftContainer('body', template.ast(`import * as React from 'react'`));
          }

          if (state.compiledImports.styled && !path.scope.getBinding('forwardRef')) {
            // forwardRef is missing - add it in at the last moment!
            path.unshiftContainer('body', template.ast(`import { forwardRef } from 'react'`));
          }

          const filename = basename(state.filename ?? '') || 'File';
          const version = process.env.TEST_PKG_VERSION || packageJson.version;

          path.addComment('leading', ` ${filename} generated by ${packageJson.name} v${version} `);

          // Add a line break after the comment
          path.unshiftContainer('body', t.noop());

          // Callback when included files have been added.
          if (this.includedFiles.length && this.opts.onIncludedFiles) {
            this.opts.onIncludedFiles(unique(this.includedFiles));
          }

          // Cleanup paths that have been marked.
          state.pathsToCleanup.forEach((clean) => {
            switch (clean.action) {
              case 'remove': {
                clean.path.remove();
                return;
              }

              case 'replace': {
                clean.path.replaceWith(t.nullLiteral());
                return;
              }

              default:
                return;
            }
          });
        },
      },
      ImportDeclaration(path, state) {
        if (path.node.source.value !== COMPILED_MODULE) {
          return;
        }

        // The presence of the module enables CSS prop
        state.compiledImports = state.compiledImports || {};

        // Go through each import and enable each found API
        path.get('specifiers').forEach((specifier) => {
          if (!state.compiledImports || !specifier.isImportSpecifier()) {
            // Bail out early
            return;
          }

          (['styled', 'ClassNames', 'css', 'keyframes'] as const).forEach((apiName) => {
            if (
              state.compiledImports &&
              t.isIdentifier(specifier.node?.imported) &&
              specifier.node?.imported.name === apiName
            ) {
              // Enable the API with the local name
              state.compiledImports[apiName] = specifier.node.local.name;
              specifier.remove();
            }
          });
        });

        if (path.node.specifiers.length === 0) {
          path.remove();
        }
      },
      TaggedTemplateExpression(path, state) {
        if (
          isCompiledCSSTaggedTemplateExpression(path.node, state) ||
          isCompiledKeyframesTaggedTemplateExpression(path.node, state)
        ) {
          state.pathsToCleanup.push({ path, action: 'replace' });
          return;
        }

        if (isCompiledStyledTaggedTemplateExpression(path.node, state)) {
          visitStyledPath(path, { context: 'root', state, parentPath: path });
          return;
        }
      },
      CallExpression(path, state) {
        if (
          isCompiledCSSCallExpression(path.node, state) ||
          isCompiledKeyframesCallExpression(path.node, state)
        ) {
          state.pathsToCleanup.push({ path, action: 'replace' });
          return;
        }

        if (isCompiledStyledCallExpression(path.node, state)) {
          visitStyledPath(path, { context: 'root', state, parentPath: path });
          return;
        }
      },
      JSXElement(path, state) {
        if (!state.compiledImports?.ClassNames) {
          return;
        }

        visitClassNamesPath(path, { context: 'root', state, parentPath: path });
      },
      JSXOpeningElement(path, state) {
        if (!state.compiledImports) {
          return;
        }

        visitCssPropPath(path, { context: 'root', state, parentPath: path });
      },
    },
  };
});
