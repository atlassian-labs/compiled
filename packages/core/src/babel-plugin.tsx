module.exports = require('@compiled/babel-plugin');
