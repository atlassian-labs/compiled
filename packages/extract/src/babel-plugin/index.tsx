import { declare } from '@babel/helper-plugin-utils';
import * as t from '@babel/types';
import { NodePath } from '@babel/core';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkgJson = require('../../package.json');

const isCreateElement = (node: t.Node): node is t.CallExpression => {
  return (
    t.isMemberExpression(node) &&
    t.isIdentifier(node.object) &&
    node.object.name === 'React' &&
    t.isIdentifier(node.property) &&
    node.property.name === 'createElement'
  );
};

const removeStyleDeclarations = (node: t.Node, parentPath: NodePath<any>) => {
  if (t.isCallExpression(node) && isCreateElement(node.callee)) {
    // We've found something that looks like React.createElement(CS)
    const children = node.arguments[2];
    if (t.isArrayExpression(children)) {
      children.elements.forEach((value) => {
        if (!t.isIdentifier(value)) {
          return;
        }

        const binding = parentPath.scope.getBinding(value.name);
        binding?.path.remove();
      });
    }
  } else if (
    t.isJSXElement(node) &&
    t.isJSXIdentifier(node.openingElement.name) &&
    node.openingElement.name.name === 'CS'
  ) {
    // We've found something that looks like <CS>
    const children = node.children[0];
    if (t.isJSXExpressionContainer(children) && t.isArrayExpression(children.expression)) {
      children.expression.elements.forEach((value) => {
        if (!t.isIdentifier(value)) {
          return;
        }

        const binding = parentPath.scope.getBinding(value.name);
        binding?.path.remove();
      });
    }
  }
};

export default declare((api) => {
  api.assertVersion(7);

  return {
    name: `${pkgJson.name}/babel-plugin`,
    visitor: {
      ImportSpecifier(path) {
        if (t.isIdentifier(path.node.imported) && ['CC', 'CS'].includes(path.node.imported.name)) {
          path.remove();
        }
      },
      CallExpression(path) {
        const callee = path.node.callee;
        if (isCreateElement(callee)) {
          // We've found something that looks like React.createElement(...)
          // Now we want to check if it's from the Compiled Runtime and if it is - replace with its children.
          const component = path.node.arguments[0];
          if (!t.isIdentifier(component) || component.name !== 'CC') {
            return;
          }

          const [, , compiledStyles, nodeToReplace] = path.get('arguments');

          // Before we replace this node with its children we need to go through and remove all the
          // style declarations from the CS call.
          removeStyleDeclarations(compiledStyles.node, path);

          // All done! Let's replace this node with the user land child.
          path.replaceWith(nodeToReplace);
          path.node.leadingComments = null;
          return;
        }

        if (t.isIdentifier(callee) && callee.name === '_jsxs') {
          // We've found something that looks like _jsxs(...)
          // Now we want to check if it's from the Compiled Runtime and if it is - replace with its children.
          const component = path.node.arguments[0];
          if (!t.isIdentifier(component) || component.name !== 'CC') {
            return;
          }

          const [, props] = path.get('arguments');
          if (!t.isObjectExpression(props.node)) {
            return;
          }

          const children = props.node.properties.find((prop): prop is t.ObjectProperty => {
            return (
              t.isObjectProperty(prop) && t.isIdentifier(prop.key) && prop.key.name === 'children'
            );
          });

          if (!children || !t.isArrayExpression(children.value)) {
            return;
          }

          const [compiledStyles, nodeToReplace] = children.value.elements as t.Expression[];

          // Before we replace this node with its children we need to go through and remove all the
          // style declarations from the CS call.
          removeStyleDeclarations(compiledStyles, path);

          // All done! Let's replace this node with the user land child.
          path.replaceWith(nodeToReplace);
          path.node.leadingComments = null;
          return;
        }
      },
    },
  };
});
