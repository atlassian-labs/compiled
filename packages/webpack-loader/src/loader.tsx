import path from 'path';
import { transformAsync } from '@compiled/babel-plugin';
import { createSetupError } from './utils/create-error';
import type { LoaderOptions } from './types';

/**
 * Compiled webpack loader.
 *
 * When used by itself it improves the local developer experience when imports used in your
 * CSS have been inlined.
 *
 * Also enables CSS extraction when the `extract` option is set to `true`.
 *
 * @param this
 * @param code
 */
export default async function compiledLoader(this: any, code: string): Promise<void> {
  const callback = this.async();

  // Bail early if Compiled isn't in the module.
  if (code.indexOf('@compiled/react') === -1) {
    return callback(null, code);
  }

  try {
    const { __sheetStore, ...options }: LoaderOptions = this.getOptions({
      type: 'object',
      properties: {
        extract: {
          type: 'boolean',
        },
        importReact: {
          type: 'boolean',
        },
        nonce: {
          type: 'string',
        },
      },
    });

    if (options.extract && !__sheetStore) {
      throw createSetupError(`The CompiledExtractPlugin has not been configured, import the plugin and add it to the plugins array in your webpack config.

  const { CompiledExtractPlugin } = require('@compiled/webpack-loader');

  plugins: [
    new HtmlWebpackPlugin(),
    new CompiledExtractPlugin(),
  ]`);
    }

    const result = await transformAsync(code, {
      filename: this.resourcePath,
      opts: { ...options, cache: true },
    });

    result.includedFiles.forEach((file) => {
      // Webpack needs to be told of any files that have been inlined into userland CSS
      // so that things get rebuilt without needing to modify the owning file.
      // Without it - you'd need to modify that file before your changes are reflected!
      this.addDependency(path.normalize(file));
    });

    if (options.extract && __sheetStore) {
      __sheetStore.add(result.sheets);
    }

    callback(null, result.code);
  } catch (e) {
    callback(e);
  }
}
