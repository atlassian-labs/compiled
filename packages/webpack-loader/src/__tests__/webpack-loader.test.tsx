import loader from '../index';
import { transformAsync } from '@compiled/babel-plugin';

jest.mock('@compiled/babel-plugin');

const getLoader = (opts: { addDependency: jest.Mock; callback: jest.Mock }) =>
  loader.bind({
    async: () => opts.callback,
    resourcePath: '/projects/index.js',
    addDependency: opts.addDependency,
    getOptions: () => ({ importReact: false }),
  });

describe('webpack loader', () => {
  it('should bail out early if the file has no compiled in it', () => {
    const loader = getLoader({ addDependency: jest.fn(), callback: jest.fn() });

    loader('console.log(undefined);');

    expect(transformAsync).not.toHaveBeenCalled();
  });

  it('should transform code if compiled has been found', () => {
    const loader = getLoader({ addDependency: jest.fn(), callback: jest.fn() });
    const code = `import '@compiled/react';`;

    loader(code);

    expect(transformAsync).toHaveBeenCalledWith(code, {
      // Filename needed for module traversal
      filename: '/projects/index.js',
      opts: {
        // Force caching module traversal calls
        cache: true,
        // Userland configuration
        importReact: false,
      },
    });
  });

  it('should callback with transformed code', async () => {
    const callback = jest.fn();
    (transformAsync as jest.Mock).mockReturnValue(
      Promise.resolve({ code: 'transformed-code', includedFiles: [] })
    );
    const loader = getLoader({ addDependency: jest.fn(), callback });

    await loader(`import '@compiled/react';`);

    expect(callback).toHaveBeenCalledWith(null, 'transformed-code');
  });

  it('should add dependencies for included files', async () => {
    const addDependency = jest.fn();
    (transformAsync as jest.Mock).mockReturnValue(
      Promise.resolve({ code: 'transformed-code', includedFiles: ['one', 'two'] })
    );
    const loader = getLoader({ callback: jest.fn(), addDependency });

    await loader(`import '@compiled/react';`);

    expect(addDependency).toHaveBeenCalledWith('one');
    expect(addDependency).toHaveBeenCalledWith('two');
    expect(addDependency).toHaveBeenCalledTimes(2);
  });

  it('should callback on error', async () => {
    const callback = jest.fn();
    (transformAsync as jest.Mock).mockReturnValue(Promise.reject('error occurred'));
    const loader = getLoader({ addDependency: jest.fn(), callback });

    await loader(`import '@compiled/react';`);

    expect(callback).toHaveBeenCalledWith('error occurred');
  });
});
