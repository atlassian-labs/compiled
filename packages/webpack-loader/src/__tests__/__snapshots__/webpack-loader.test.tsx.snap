// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack loader should callback on error 1`] = `
[SyntaxError: /projects/index.js: A numeric separator is only allowed between two digits (1:29)

[0m[31m[1m>[22m[39m[90m 1 |[39m [36mimport[39m [32m'@compiled/react'[39m[33m;[39m [35m222[39m_[33m![39m[33m![39m[33m![39m[0m
[0m [90m   |[39m                              [31m[1m^[22m[39m[0m]
`;
